pkg/apis/metal3/v1alpha1/baremetalhost_types.go

```diff
@@ -641,15 +638,15 @@ func (host *BareMetalHost) WasProvisioned() bool {
 // NeedsDeprovisioning compares the settings with the provisioning
 // status and returns true when the host should be deprovisioned.
 func (host *BareMetalHost) NeedsDeprovisioning() bool {
-	if host.Spec.ExternallyProvisioned {
-		return false
+	if host.Spec.Image == nil {
```

Comment:
This changes the meaning of this method. Before we had the higher-level state machine, this used to be trying to indicate when a transition needed to happen to clean a host. Now it means something closer to "should this host not have an image on it" and leaving it up to the caller to decide if that means a transition is needed. Maybe we should rename the method?

Reply:
Yeah, it still has that meaning but only when called from certain states. I'm open to suggestions for a better name... `IsProvisionedOrProvisioningAndNeedsDeprovisioning()` didn't quite trip off the tongue ;)
Another option is to check the state inside this function so that it doesn't return true if provisioning hasn't at least started. I was trying to avoid making these functions state-aware and keep the state logic in the state machine. Probably better that it be explicit rather than you have to remember where it's safe to call it from though.
Maybe this function should just be moved to the state machine and made private instead of living here and inviting people to call it for more general purposes?

Reply:
I opened a separate PR #397 to address this, since this one was blocking people and got merged already.